### PR TITLE
feat: add molecule of checkpoint, metadata, stake, delegate, reward

### PR DESCRIPTION
**Pull request by wenyuanhust**

### CODE REVIEW

File: common/types/schemas/delegate.mol [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import basic;
+
+table DelegateLimit {
+   threshold:          Uint128,
+   max_delegator_size: Uint32,
+}
+
+table DelegateInfo {
+    l1_address:     Identity,
+    l2_address:     Identity,
```
Comment (felicityin) [inline common/types/schemas/delegate.mol]: l2_address is useless.

File: common/types/schemas/stake.mol [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    stake_smt_type_id: Byte32,
+    staker_identity:   Identity,
+}
+
+// The data[0..16] is a le_u128 means xUDT amount.
+// So the cell data is decode from offset 16.
+table StakeAtLockCellData {
+    version:            byte,
+    l1_address:         Identity,
+    l2_address:         Identity,
```
Comment (felicityin) [inline common/types/schemas/stake.mol:19]: Useless fields.

File: common/types/schemas/metadata.mol [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    address:       Identity,
+    propose_count: Uint64,
+}
+
+vector MetadataList <Metadata>;
+vector ValidatorList <Validator>;
+vector ValidatorHistoryList <ValidatorHistory>;
+
+table Metadata {
+    epoch_len:       Uint32, //how many periods as one epoch
```
Comment (felicityin) [inline common/types/schemas/metadata.mol]:
```suggestion
    epoch_len:       Uint32, // how many periods as one epoch
```

File: common/types/schemas/metadata.mol [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    period_len:      Uint32, //how many blocks as one period
```
Comment (felicityin) [inline common/types/schemas/metadata.mol]:
```suggestion
    period_len:      Uint32, // how many blocks as one period
```

File: common/types/schemas/checkpoint.mol [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+table CheckpointLockCellData {
+    version:           byte,
+    epoch:             Uint64,
+    period:            Uint32,
+    total_period:      Uint32,
+    state_root:        Byte32,
+    latest_block_hash: Byte32,
+    xudt_type_id:      Byte32,
+    timestamp:         Uint64,
```
Comment (felicityin) [inline common/types/schemas/checkpoint.mol:18]: block or height

File: common/types/schemas/checkpoint.mol [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    count:   Uint32,
+}
+
+vector ProposeCounts <ProposeCount>;
+
+table CheckpointLockCellData {
+    version:           byte,
+    epoch:             Uint64,
+    period:            Uint32,
+    total_period:      Uint32,
```
Comment (felicityin) [inline common/types/schemas/checkpoint.mol:14]: Useless field.

File: common/types/schemas/delegate.mol [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import basic;
+
+table DelegateLimit {
+   threshold:          Uint128,
+   max_delegator_size: Uint32,
```
Comment (felicityin) [inline common/types/schemas/delegate.mol]:
```suggestion
    threshold:          Uint128,
    maximum_delegators: Uint32,
```

File: common/types/schemas/checkpoint.mol [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    state_root:        Byte32,
+    latest_block_hash: Byte32,
+    xudt_type_id:      Byte32,
+    timestamp:         Uint64,
+    propose_count:     ProposeCounts,
+}
+
+table CheckpointLockWitnessLock {
+	proposal: Bytes,
+	proof:    Bytes,
```
Comment (felicityin) [inline common/types/schemas/checkpoint.mol]:
```suggestion
    proposal: Bytes,
    proof:    Bytes,
```

File: common/types/schemas/checkpoint.mol (hunk 6 above)
Comment (felicityin) [inline common/types/schemas/checkpoint.mol]: Useless field.

File: common/types/schemas/checkpoint.mol (hunk 5 above)
Comment (felicityin) [inline common/types/schemas/checkpoint.mol]: block or height

File: common/types/schemas/stake.mol [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+table StakeLockArgs {
+    stake_smt_type_id: Byte32,
+    staker_identity:   Identity,
+}
+
+// The data[0..16] is a le_u128 means xUDT amount.
+// So the cell data is decode from offset 16.
+table StakeAtLockCellData {
+    version:            byte,
+    l1_address:         Identity,
```
Comment (felicityin) [inline common/types/schemas/stake.mol:18]: Useless fields.

File: common/types/schemas/stake.mol (hunk 2 above)
Comment (felicityin) [inline common/types/schemas/stake.mol:19]: Useless field.

File: common/types/schemas/delegate.mol [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+table DelegateInfo {
+    l1_address:     Identity,
+    l2_address:     Identity,
+    dividend_ratio: byte, // only dividend_ratio% reward will go to delegators
+}
+
+table DelegateCellData {
+    version:            byte,
+    delegate_limit:     DelegateLimit,
+    delegate_info:      DelegateInfo,
```
Comment (felicityin) [inline common/types/schemas/delegate.mol:17]: Why need this?

File: common/types/schemas/delegate.mol [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import basic;
+
+table DelegateLimit {
+   threshold:          Uint128,
+   max_delegator_size: Uint32,
+}
+
+table DelegateInfo {
+    l1_address:     Identity,
```
Comment (felicityin) [inline common/types/schemas/delegate.mol]: staker or staker_l1_address

**Review by KaoImin — APPROVED** [review]

File: common/types/schemas/checkpoint.mol (hunk 5 above)
Comment (KaoImin) [reply common/types/schemas/checkpoint.mol]: The Axon block number can be calculated from epoch and period, this is the timestamp in the corresponding Axon block header.

File: common/types/schemas/checkpoint.mol (hunk 6 above)
Comment (wenyuanhust) [reply common/types/schemas/checkpoint.mol]: total_period may be useless, why others? where can we store epoch and period of this checkpoint

File: common/types/schemas/stake.mol (hunk 2 above)
Comment (wenyuanhust) [reply common/types/schemas/stake.mol:19]: can not recover address just from lock args

File: common/types/schemas/checkpoint.mol (hunk 6 above)
Comment (felicityin) [reply common/types/schemas/checkpoint.mol]: Yes, only total_period is useless. I don't know why it references multi lines.

File: common/types/schemas/stake.mol (hunk 2 above)
Comment (felicityin) [reply common/types/schemas/stake.mol:19]: l2_address is useless.

**Pull request merged**